### PR TITLE
[#IC-50] Composed Cosmos Versioned Model

### PR DESCRIPTION
**Pull request by BurnedMarshal**
#### List of Changes

A new model type is created `CosmosdbModelComposedVersioned`.
Refactory of `CosmosdbModelVersioned` to allow override of `createNewVersion` method.

#### Motivation and Context

We need a new model type that handle versioning for documents with a couple externalKey-partitionKey.
This model works correctly with a limited range of externalKey for a specific partition key.

#### How Has This Been Tested?

unit test

#### Types of changes

- [ ] Chore (nothing changes by a user perspective)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### CODE REVIEW

**Review by balanza — COMMENTED** [review]
> This model works correctly with a limited range of externalKey for a specific partition key

Can you explain this a bit more?

---

Couldn't we generalize `CosmosDbVersioneModel` so that it can handle both simply keys (scalar `modelId`) and composed keys (pair `[modelId, partititonKey]`)? By finding a criteria and by overloading `generateVersionedModelId` we can achieve this. I don't actually know if it's a big value anyway, but I feel like we're really doing the same things in both classes.

Also, consider to write some integrations tests, as they may be more eloquent in terms of what this class is intended for.

**Comment by BurnedMarshal** [comment]
> > This model works correctly with a limited range of externalKey for a specific partition key
> 
> Can you explain this a bit more?

If documents with different external keys grows inside **the same** partition key, the queries will be more expensive.
Nevertheless a composed index including external key and partition key is required on all the models that will use this data structure.

> 
> Couldn't we generalize `CosmosDbVersioneModel` so that it can handle both simply keys (scalar `modelId`) and composed keys (pair `[modelId, partititonKey]`)? By finding a criteria and by overloading `generateVersionedModelId` we can achieve this. I don't actually know if it's a big value anyway, but I feel like we're really doing the same things in both classes.
> 

That was my fist idea, but we already have a Model that has different modelId and partitionKey and the signature we needed for `generateVersionedModelId` it's different in our case. The combination of this two reason forced me to implement this abstract model.

> Also, consider to write some integrations tests, as they may be more eloquent in terms of what this class is intended for.

It's my intention do it on the implementation of this class (es. ActivationModel).

File: src/utils/__tests__/cosmosdb_model_composed_versioned.test.ts [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+    if (E.isLeft(result)) {
+      expect(result.left.kind).toBe("COSMOS_ERROR_RESPONSE");
+      if (result.left.kind === "COSMOS_ERROR_RESPONSE") {
+        expect(result.left.error.code).toBe(500);
+      }
+    }
+  });
+});
+
+describe("update", () => {
```
Comment (michaeldisaro) [inline src/utils/__tests__/cosmosdb_model_composed_versioned.test.ts:162]: I think this should be "create"

File: src/utils/__tests__/cosmosdb_model_composed_versioned.test.ts [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+        ...aMyDocument,
+        ...someMetadata,
+        id: documentId(externalKey, pk, expectedNextVersion),
+        version: expectedNextVersion
+      });
+    }
+  });
+});
+
+describe("create", () => {
```
Comment (michaeldisaro) [inline src/utils/__tests__/cosmosdb_model_composed_versioned.test.ts:191]: I think this should be "update"

File: src/utils/__tests__/cosmosdb_model_composed_versioned.test.ts [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+  it("should create a new document with version 0", async () => {
```
Comment (michaeldisaro) [inline src/utils/__tests__/cosmosdb_model_composed_versioned.test.ts:192]: This title seems wrong. From the test I understand that you are "updating" a retrieved document, and this should create a new version as defined in row 193.

**Review by michaeldisaro — COMMENTED** [review]
I have some doubts on tests.

File: src/utils/__tests__/cosmosdb_model_composed_versioned.test.ts (hunk 1 above)
Comment (BurnedMarshal) [reply src/utils/__tests__/cosmosdb_model_composed_versioned.test.ts]: fixed in 8532fec

File: src/utils/__tests__/cosmosdb_model_composed_versioned.test.ts (hunk 2 above)
Comment (BurnedMarshal) [reply src/utils/__tests__/cosmosdb_model_composed_versioned.test.ts]: fixed in 8532fec

File: src/utils/__tests__/cosmosdb_model_composed_versioned.test.ts (hunk 3 above)
Comment (BurnedMarshal) [reply src/utils/__tests__/cosmosdb_model_composed_versioned.test.ts]: Yes, the titles of the unit test was not corresponding with the scenario. Fixed in 8532fec
Thanks

**Review by michaeldisaro — APPROVED** [review]
LGTM

**Comment by balanza** [comment]
My first thought when I red `CosmosdbModelComposedVersioned` was we were versioning two entities at the same time. Then I looked at the code and I realized that was about having a composite identity for the document. 

I does not occur to me a more eloquent name for the class, nor I think I should nitpick too much on naming. Nevertheless, as this is a base class we may rename it in future with little effort.

My advice is to ad an eloquent doc block describing the purpose of the class.

For the rest: I'm fine with the code, good job!

**Comment by BurnedMarshal** [comment]
> My first thought when I red `CosmosdbModelComposedVersioned` was we were versioning two entities at the same time. Then I looked at the code and I realized that was about having a composite identity for the document.
> 
> I does not occur to me a more eloquent name for the class, nor I think I should nitpick too much on naming. Nevertheless, as this is a base class we may rename it in future with little effort.
> 
> **My advice is to add an eloquent doc block describing the purpose of the class.**
> 
> For the rest: I'm fine with the code, good job!

@balanza I've added some comments in 80be35d

**Pull request merged**